lib/active_merchant/billing/gateways/checkout_v2.rb

```diff
@@ -81,43 +82,59 @@ def scrub(transcript)
-        post[:card][:expiryMonth] = format(payment_method.month, :two_digits)
+        post[:source] = {}
+        post[:source][:type] = 'card'
+        post[:source][:name] = payment_method.name
+        post[:source][:number] = payment_method.number
+        post[:source][:cvv] = payment_method.verification_value
+        post[:source][:expiry_year] = format(payment_method.year, :four_digits)
+        post[:source][:expiry_month] = format(payment_method.month, :two_digits)
       end
 
       def add_customer_data(post, options)
-        post[:email] = options[:email] || 'unspecified@example.com'
-        post[:customerIp] = options[:ip] if options[:ip]
+        post[:customer] = {}
+        post[:customer][:email] = options[:email] || nil
+        post[:payment_ip] = options[:ip] if options[:ip]
         address = options[:billing_address]
-        if(address && post[:card])
-          post[:card][:billingDetails] = {}
-          post[:card][:billingDetails][:addressLine1] = address[:address1]
-          post[:card][:billingDetails][:addressLine2] = address[:address2]
-          post[:card][:billingDetails][:city] = address[:city]
-          post[:card][:billingDetails][:state] = address[:state]
-          post[:card][:billingDetails][:country] = address[:country]
-          post[:card][:billingDetails][:postcode] = address[:zip]
-          post[:card][:billingDetails][:phone] = { number: address[:phone] } unless address[:phone].blank?
+        if(address && post[:source])
+          post[:source][:billing_address] = {}
+          post[:source][:billing_address][:address_line1] = address[:address1] if address[:address1]
+          post[:source][:billing_address][:address_line2] = address[:address2] if address[:address2]
+          post[:source][:billing_address][:city] = address[:city] if address[:city]
+          post[:source][:billing_address][:state] = address[:state] if address[:state]
+          post[:source][:billing_address][:country] = address[:country] if address[:country]
+          post[:source][:billing_address][:zip] = address[:zip] if address[:zip]
+          post[:source][:phone] = { number: address[:phone] } unless address[:phone].blank?
         end
       end
 
       def add_transaction_data(post, options={})
-        post[:cardOnFile] = true if options[:card_on_file] == true
-        post[:transactionIndicator] = options[:transaction_indicator] || 1
-        post[:previousChargeId] = options[:previous_charge_id] if options[:previous_charge_id]
+        post[:card_on_file] = true if options[:card_on_file] == true
+        post[:payment_type] = 'Regular' if options[:transaction_indicator] == 1
+        post[:payment_type] = 'Recurring' if options[:transaction_indicator] == 2
+        post[:previous_payment_id] = options[:previous_charge_id] if options[:previous_charge_id]
+      end
+
+      def add_3ds(post, options)
+        return unless options[:execute_threed] || options[:threed_dynamic]
```

Comment:
Since you don't really use `threed_dynamic` value, I'd just drop its dependency. Also, if we can rely on the presence of `three_d_secure` key, would be a bit more aligned with other gateways that already support the param pass thru.
```ruby
{
  three_d_secure: {
    cavv: 'cavv value goes here',
    eci: 'eci value',
    xid: 'xid value',
  }
}
```

then this method would be something like
```suggestion
def add_3ds(post, options)
  if options[:three_d_secure]
    post[:'3ds'] = {}
    post[:'3ds'][:enabled] = true
    post[:'3ds'][:eci] = options[:eci] if options[:eci]
    post[:'3ds'][:cryptogram] = options[:cavv] if options[:cavv]
    post[:'3ds'][:xid] = options[:xid] if options[:xid]
  end
end
```

Reply:
a `return unless options[:three_d_secure]` would work too